### DISCUSSION TITLE
Creality Halot Mage - Export to *.cxdlpv4

### DISCUSSION THREAD

**Discussion by codeppl** [sn4k3/UVtools#714] [Q&A]
I just got a Halot Mage and am unable to edit/clear islands using the Chitubox as effectively as UV tools! unfortunately the format - *.cxdlpv4 is not supported to open/export in UV tools. And the printer won't accept any other format other than *.cxdlpv4!

Request to share any info on possibility to integrate the format "*.cxdlpv4" with UV tools or any alternate method to treat the *.cxdlpv4 format through UV Tools.

Thanks in advance!

**Comment by sn4k3** [comment]
It takes .CTB too, export as CTB as it will open on UVtools.

**Reply by sartymlg** [reply]
The issue is, we cannot export as cxdlpv4 in order to print it, before perform any fix.

Congrats for your job! Is my first resin printer, and your app is simply amazing.

**Reply by sn4k3** [reply]
Yes, you need to export as CTB, your printer accepts it and probably will give you more features.

**Reply by sartymlg** [reply]
I think mage has a custom motherboard, not a chitubox one, so probably doesn't support ctb. I'm not sure, but I read many comments complaing that lechee and chitubox wasn't compatible with it yet and only can be used the creality software  (I discovered that chitubox pro supports it).

**Reply by sn4k3** [reply]
Even still, try it ;) CTB encrypted version. You can use my PrusaSlicer profile for Mage that will do that.

**Reply by kenweise** [reply]
I have tried CTB, CTB Encrypted, and others. My Mage Pro will not read anything but cxdlpv4.

**Reply by sartymlg** [reply]
Hi kenweise, maybe this can help you. I didn't tried exporting the file with us tools, but i did q quick test with lechee slicer configuring it as a Saturn 2 and exporting in ctb latest version and my mage 8k (not pro) read it, miniature was displayed and started to print, after that I cancelled the print, but seems that it could read it properly.

**Reply by kenweise** [reply]
I tried CTB exported from UVTools. The Mage Pro would not open it.

**Reply by sn4k3** [reply]
> I tried CTB exported from UVTools. The Mage Pro would not open it.

You need the Encrypted CTB, not CTB

**Reply by kenweise** [reply]
It will not open he encrypted CTB. It is shown in the file list, but says it is an invalid file when I try to open it.

**Reply by sn4k3** [reply]
Then there are no way to use UVtools yet with that printer

**Comment by kenweise** [comment]
Is there a way to convert this thread to a feature request?

User Story: As a Halot Mage Pro user, I would like to use UVtools to validate my models and calibrate the printer.

Requirement: Capability to import from and convert to the CXDLPV4 file format. 

Acceptance criteria
- UVTools can import existing CXDLPV4 files.
- UVTools can convert files to the CXDLPV4.
- Users can load and print the converted files on their Mage or Mage Pro printer

**Reply by sn4k3** [reply]
No, I can only create issue and link the discussion, but in any case, it's not necessary.
I'm aware of the format, it's just they are never given and it takes time to dig and discover how they are read.
I already start working on it but I can't give any time expectation.

**Reply by ememolero** [reply]
I second the motion of @kenweise 
I will do a money donation for that if it helps to advance it as fast as you can!! Thanks

**Reply by ememolero** [reply]
You did it! Thank you! But the Printer didn't recognize the file. It throws the error: "This model does not match the printer". I can import the model to the printer, it makes a check and it's readable. But when I push the Play button it throws the previous error. 

I am attaching a simple model generated with Chitubox without errors that is recognized by the Halot Mage Pro 8K printer. From what I see in UV Tools I can change many parameters of LightOff time, up and down speeds that neither the original Creality slicer nor the Chitubox Plugin allow for this format. From what I have read and tested, it seems that the printer has limitations when it comes to varying certain parameters, it does not support waiting times longer than 20 sec or speeds lower than 60mm/min. Stupid things.

By the way, I'm not very familiar with using github and I just discovered the sponsors program ;)
[Cuadrado_Prueba_1ud.zip](https://github.com/sn4k3/UVtools/files/11910254/Cuadrado_Prueba_1ud.zip)

**Reply by sn4k3** [reply]
I think I might have seen the problem, try this file processed with UVtools, does it work?
[Cuadrado_Prueba_1ud_01h13m_2ml_UVtools.zip](https://github.com/sn4k3/UVtools/files/11910336/Cuadrado_Prueba_1ud_01h13m_2ml_UVtools.zip)

**Reply by ememolero** [reply]
Thank you!. It works, I can print as is. But, if I open it with UVTools and save again with or without modifications in parameters or some painting in any layer the file stop working with the infamous error. Now I'm doing a print test with that file you sent my as is.

**Reply by ememolero** [reply]
Apparently everything works... he prints fine, I have to see the final result. Other considerations to take into account for the parameters:

- The printer has an internal parameter editor that can be chosen to print instead of the parameters that come in the file.

- The printer has 3 operating modes, Normal, Dynax and Dynax+. Normal operation allows fewer modifications in the parameters, only those that appear in the first image. Dynax modes unlock more parameters to modify (second image). Something called "cushion". This is to set an initial print height with other speed and curing parameters, something like managing the bottom layers separately up to a defined height and then having it take the parameters for normal layers. Currently I use the Dynax mode which is the closest thing to two-stage printing, however taking off the platform at 1mm/s (minimum possible supported) is an aberration:

NORMAL MODE:
![1 NORMAL MODE](https://github.com/sn4k3/UVtools/assets/37576277/5094b9a4-ffed-4d31-964f-6da8f39e49a4)

WITH DYNAX MODE ACTIVATED:

![2 WITH DYNAX MODE](https://github.com/sn4k3/UVtools/assets/37576277/6c3e9d59-1a9a-48d3-8b0c-8a7dc2fd033d)


- In normal mode the light off delay for bottom and normal layers is defined by the setting LIGHT OFF DELAY. With Dynax modes, the light off delay for bottom layer is defined by the setting "CUSHION LAYER TURN OFF DELAY" 

- The only parameter that can be varied in steps of 0.1 is the exposure time of the normal layers, the rest are in steps of 1

- As the names of the parameters are in motion, in some photos the name of the parameter is not visible at all so I have written some equivalent references that we already know as bottom layer or light off so that it is not very confusing.

I also attach several screenshots with the maximum and minimum values for each parameter recognized by the printer:

![3 MIN_SETTINGS](https://github.com/sn4k3/UVtools/assets/37576277/faa4a9e4-eaa7-41ad-ab18-8a3a925b8673)

![4 MIN_SETTINGS (Cushion Part)](https://github.com/sn4k3/UVtools/assets/37576277/a508daf2-57ee-4eea-ac1e-8a4408c087ca)

MAX SETTINGS:

![5 MAX_SETTINGS](https://github.com/sn4k3/UVtools/assets/37576277/08738dd4-f302-41d0-b8de-22acf6de2f6d)
![6 MAX_SETTINGS (Cushion Part)](https://github.com/sn4k3/UVtools/assets/37576277/45fcb27c-406b-4aa7-93b6-8555b32df8ad)


I hope all this helps you with this madness. I still don't understand why they are unable to lower the speed to less than 60mm/min, I don't know if it will be because of a decimal point issue. I have programmed electronics for simple printers that I have built myself and I did not have any problems. I'm planning to bypass the stepper motor with Arduino as a stepper divider since the motor is configurable with microswitches and try to lower the speed by half.

Thank you.

**Reply by ememolero** [reply]
Another Issue with the file you sent me. The first time I Started the print, The platform went down as usual but after a few seconds in the bottom, the printer gave the job as finished without any warning or text. However, if I print the job again, it starts without problems and print layers, I have not yet been able to see if it is complete, I am verifying it.

**Reply by sn4k3** [reply]
> I still don't understand why they are unable to lower the speed to less than 60mm/min

Because they set it to mm/s without floating point and then 1 is the min.

>  I have programmed electronics for simple printers that I have built myself and I did not have any problems

I can tell you, none of these chinese printers are good in firmware, they should be forbidden to do firmwares. They are very bad at it. Never learn from mistakes nor listen to customers. They only care about profit with junk products.

**Reply by sn4k3** [reply]
Test the new patch

**Reply by ememolero** [reply]
I'm on it!

**Reply by ememolero** [reply]
It works like a Charm, for now I will do some tests and verify everything is ok. 

Thank you so much, you are awesome!

**Reply by sn4k3** [reply]
Thanks for confirming.
Note that UVtools offer per layer settings and many other settings not visible on slicer, because file spec supports it but It doesn't mean the printer firmware will, need testing here.

**Reply by ememolero** [reply]
Ok. I understand. I think is some related the way the mainboard is programed. In fact, the speed 1mm/s is not real, I can see that it is at least 5mm/s. For now is far enough to use to apply my print mode, the use of scripts and also to be able to verify the models. If I detect anything strange in tests I will let you know. Thank you again

**Comment by kenweise** [comment]
I just tried with V3.15.0. I imported a complex model and exported an exposure test. Now I just need to run it.

**Reply by sn4k3** [reply]
Won't work as there is a bug

**Reply by codeppl** [reply]
Hi @sn4k3, first of all thanks to you for introducing the cxdlpv4 format! I am using the latest version..
I was able to open the files and see the layers properly but there seems to be problems with issues detection. it seems the file reading is having some bugs showing incorrect islands. The islands are popping on both sides of the layer and they are not from the model but from the interpretation of the file. I hope you are aware of the issue. 
Please let me know if you need any snapshots for reference.

**Reply by sn4k3** [reply]
Without a sample file and images showing the problem I have no clue. 
Most of "bugs/fake" issue detection are not that but due user thresholds that are configurable.
Also, if you move the model position on plate you need to redetect issues, ie, after apply model reposition suggestion.

**Reply by codeppl** [reply]
I have just got 10+ hours print completed! the file is directly from Chitubox! It seems the issue is within the file export from Chitubox itself! the print has flipped and merged on the same model and is on alternate layers. I thought the issue is from UV Tools but no! the file exported from chitu itself is f*d up! I will try again tomorrow and capture snaps/scrnshts for reference..

**Reply by sn4k3** [reply]
You may read this: https://www.facebook.com/groups/uvtools/posts/935295364246634

**Reply by codeppl** [reply]
Got the issue..! it was the problem with Printer settings Mirror option - "LCD_mirror". I have changed the option to Normal and it is working properly now! UVTools is displaying the layers perfectly! 
![image](https://github.com/sn4k3/UVtools/assets/9064535/e8eee0ce-d93a-4713-a5a0-36d6e9ec94f6)
